### PR TITLE
MAINT: Drop Python 3.9

### PR DESCRIPTION
**Pull request by charris**
Drop Python 3.9 from the supported Python versions and CI.
Cygwin is excepted because it does not yet support 3.10.

### CODE REVIEW

**Comment by charris** [comment]
I've disabled the Cygwin workflow because it is failing the build with the new pyproject.toml. Cygwin plans to skip 
to Python 3.12 in the near future.
```
Only python 3.8 and 3.9 are updated.
Python 2, 3.6 and 3.7 are not anymore supported upstream.
Python 3.12 will be in the near future introduced and we will skip 3.10 
and 3.11 
```

File: tools/ci/cirrus_wheels.yml [hunk 1]
```diff
@@ -23,10 +23,6 @@ linux_aarch64_task:
     # build in a matrix because building and testing all four wheels in a
     # single task takes longer than 60 mins (the default time limit for a
     # cirrus-ci task).
-    - env:
-        CIRRUS_CLONE_SUBMODULES: true
-        CIBW_BUILD: cp39-*
-        EXPECT_CPU_FEATURES: NEON NEON_FP16 NEON_VFPV4 ASIMD ASIMDHP ASIMDDP ASIMDFHM
     - env:
```
Comment (charris) [inline tools/ci/cirrus_wheels.yml:26]: Do we want to move the `EXPECT_CPU_FEATURES` line to the `3.10` test?

Comment (andyfaff) [reply tools/ci/cirrus_wheels.yml:26]: It wasn't clear to me when I was modifying the ci where these are actually used.

**Comment by ngoldbaum** [comment]
Not sure why, but it looks like the freebsd tests didn't trigger on this PR and did on #26243. Those tests will need to be updated because freebsd defaults to python3.9.

**Comment by ngoldbaum** [comment]
> Maybe pkg install -y python310 in the virtual environment?

Yeah, I think you need to explicitly install python3.10 then e.g. use a `python3.10` binary explicitly for the rest of the script. Also it looks like 3.11 is packaged, so you could use that too.

File: tools/ci/cirrus_arm.yml [hunk 2]
```diff
@@ -86,22 +87,22 @@ freebsd_test_task:
   prepare_env_script: |
     # Create a venv (the `source` command needs bash, not the default sh shell)
     chsh -s /usr/local/bin/bash
-    python -m venv .venv
+    python311 -m venv .venv
```
Comment (ngoldbaum) [inline tools/ci/cirrus_arm.yml:90]: I think there's a `.` in the name of the binary.
```suggestion
    python3.11 -m venv .venv
```

**Comment by charris** [comment]
> it looks like the freebsd tests didn't trigger on this PR

Looks like removing the [wheel build] fixes that, so there is a logic error somewhere.

**Comment by charris** [comment]
Should be good to go.

File: .github/workflows/linux.yml [hunk 3]
```diff
@@ -74,7 +74,7 @@ jobs:
         fetch-tags: true
     - uses: actions/setup-python@82c7e631bb3cdc910f68e0081d67478d79c6982d # v5.1.0
       with:
-        python-version: 'pypy3.9-v7.3.12'
+        python-version: 'pypy3.10-v7.3.12'
```
Comment (mattip) [inline .github/workflows/linux.yml:77]:
```suggestion
        python-version: 'pypy3.10-v7.3.15'
```

File: building_with_meson.md [hunk 4]
```diff
@@ -1,7 +1,7 @@
 # Building with Meson
 
 _Note: this is for early adopters. It has been tested on Linux and macOS, and
-with Python 3.9-3.12. Windows will be tested soon. There is one CI job to keep
+with Python 3.10-3.12. Windows will be tested soon. There is one CI job to keep
```
Comment (mattip) [inline building_with_meson.md:4]: We definitely support windows now. Is there really only one CI job? I think the whole CI now uses meson. Maybe this whole paragraph should be deleted.
```suggestion
with Python 3.10-3.12. There is one CI job to keep
```

**Review by mattip — APPROVED** [review]
Two small nits, otherwise LGTM

**Comment by DWesl** [comment]
> I've disabled the Cygwin workflow because it is failing the build with the new pyproject.toml. Cygwin plans to skip to Python 3.12 in the near future.
> 
> >```
>> Only python 3.8 and 3.9 are updated.
>> Python 2, 3.6 and 3.7 are not anymore supported upstream.
>> Python 3.12 will be in the near future introduced and we will skip 3.10 
>> and 3.11 
>> ```
[Source for Python update plans](https://cygwin.com/pipermail/cygwin-announce/2024-February/011613.html).

[It looks like CPython dropped some Cygwin-specific sections a few years back](https://github.com/python/cpython/issues/105790#issuecomment-1607309834), and [are aware the current build system is hard to change](https://github.com/python/cpython/issues/105790#issuecomment-1717502219), so getting a new version of Python working on Cygwin might take longer than usual.

In the meantime, I think I saw mentions of CI runs on BSD, which should keep Cygwin from breaking too badly.  I think SIMD stack alignment issues would be the main difference from Linux they wouldn't notice.

For the question from #26244, I suspect dropping Cygwin will make the other platforms simpler.  I'll be back whenever a test version of 3.12 comes out to figure out how to get CI to use that (either cygwin/cygwin-install-action#12 or the version-specifications of egor-tensin/setup-python).

**Pull request merged**